Avoid cached data between plugin generation

emanuel2258 opened the pull request:
Created a private static method that re‑initializes every static field used to hold merge state and it's used to ensuring each invocation begins with a clean slate, regardless of any prior merges.


- [ ] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [ ] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [ ] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Fixes #31263

emanuel2258 commented:
!build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=0196826d-b8f2-4b90-aa6d-890b7287ec97) - :x: completed with errors/failures.)
spawn.fullfat.buckets=com.ibm.ws.webserver.plugin.utility_fat

_Note: Target locations of links might be accessible only to IBM employees._

pnicolucci reviewed (changes requested):
`static` was added here: https://github.com/OpenLiberty/open-liberty/pull/24627 we need to understand why that was. If these variables don't need to be static we should fix that. If we have test failures we need to understand why and determine a solution.

volosied commented:
I encounter failures when I change private `PluginInfo[] plugins` variable to non-static. 


A clue as to why this variable needs to be static is how the merge tool is used, via a singleton-ish pattern: 

` PluginMergeToolImpl toolInstance = new PluginMergeToolImpl();` within the `merge(String argv[])`  method. After toolInstance is created, ` plugins = new PluginInfo[fileList.length];` is made.

Another place to look at is where ` plugins = new PluginInfo[files.length - 1];` was initialized. [Before it was in the loadData method](https://github.com/OpenLiberty/open-liberty/blob/a2d78c05cf2f8700a844cb768c142e5cdf5064de/dev/com.ibm.ws.http.plugin.merge/src/com/ibm/ws/http/plugin/merge/internal/PluginMergeToolImpl.java#L455), [but now it's within merge method. 
](https://github.com/OpenLiberty/open-liberty/blame/12950c2a17e7b63dbde5c070f30769f74634a48e/dev/com.ibm.ws.http.plugin.merge/src/com/ibm/ws/http/plugin/merge/internal/PluginMergeToolImpl.java#L739)


This `plugins` object would then be accessible by toolInstance since it's static.  If it's not static, then you'll hit NPEs (`Caused by: java.lang.NullPointerException: Cannot read the array length because "this.plugins" is null`). 


I have a branch here with some local changes for testing: https://github.com/volosied/open-liberty/tree/TS018131209-vlad

The [789a331](https://github.com/volosied/open-liberty/commit/789a331c96cb10156e641be47497e0809092b35a) commit has all tests passing while the [c9fc73d](https://github.com/volosied/open-liberty/commit/c9fc73d7bc32830ca85a3d92ac726fc6357d958c) commit has only the `testEndpointTest3` test failing.

volosied commented on dev/com.ibm.ws.http.plugin.merge/src/com/ibm/ws/http/plugin/merge/internal/PluginMergeToolImpl.java:
```diff
@@ -129,12 +126,29 @@ private void debug(String content) {
+        isXdOnly = true;
+        debug = false;
+        seqNum = 0;
+        precedence = false;
```
precedence was static before -- might be better to not reset it?

emanuel2258 replied:
Good point. I can remove it from the reset if we go this route.

emanuel2258 committed:
Turning static variables to instance scoped variables. Removed the reset static State method.

emanuel2258 commented:
I made a new commit to the PR. I reverted a lot of the static variables to instance state instead of resetting the state as a work around.  When I say reverted I mean compared to what it was prior to PR https://github.com/OpenLiberty/open-liberty/pull/24627. There is only one variable that was left static `sharedPlugins`. If I modified it to non-static it would cause problems in one of the tests in `com.ibm.ws.webserver.plugin.utility_fat`, so had the potential to regress. But if I leave it as static it still causes problems for the customer’s issue because I would still get cached data. In order to get around this I had to add  `sharedPlugins = new ArrayList<PluginInfo>();` at the beginning of the `merge` function. Please take a look when you can and let me know if you want me to modify anything else.

pnicolucci commented on dev/com.ibm.ws.http.plugin.merge/src/com/ibm/ws/http/plugin/merge/internal/PluginMergeToolImpl.java:
```diff
@@ -716,30 +718,36 @@ private boolean validateEach(String [] inputFiles, String mergeFileName) {
-        fileList = filesList.toArray(fileList);
-        plugins = new PluginInfo[fileList.length];
+        String[] fileList = filesList.toArray(new String[filesList.size()]);
+        // plugins = new PluginInfo[fileList.length];
```
Remove if not needed

pnicolucci commented on dev/com.ibm.ws.http.plugin.merge/src/com/ibm/ws/http/plugin/merge/internal/PluginMergeToolImpl.java:
```diff
@@ -221,8 +223,41 @@ private void printMergedCopy(String output) throws IOException, ParserConfigurat
+
+
+   /*     Node[] vhgs = null;
+        itrSharedPlugins = sharedPlugins.iterator();
```
Can this commented out code be removed?

pnicolucci reviewed (changes requested):
I'd like to see a test added as well

pnicolucci commented on dev/com.ibm.ws.http.plugin.merge/src/com/ibm/ws/http/plugin/merge/internal/PluginMergeToolImpl.java:
```diff
@@ -188,8 +190,8 @@ private void printMergedCopy(String output) throws IOException, ParserConfigurat
-        final Comment comment = mergeDoc.createComment(" This config file was generated by pluginUtility merge v1.0.73 on " +
-                                                       dateTimeFormatter.format(ZonedDateTime.now()) + " ");
+        final Comment comment = mergeDoc.createComment(" This config file was generated by pluginUtility merge v1.0.75 on " +
+                                                       new SimpleDateFormat("yyyy.MM.dd 'at' HH:mm:ss z").format(new Date()) + " ");
```
Why did we revert this? https://github.com/OpenLiberty/open-liberty/commit/a0d92abfc51c367a89d0e680d7c2ceac697c4ee2

emanuel2258 committed:
Incorporating code change after peer review.

pnicolucci commented on dev/com.ibm.ws.http.plugin.merge/src/com/ibm/ws/http/plugin/merge/internal/PluginMergeToolImpl.java:
```diff
@@ -188,7 +192,7 @@ private void printMergedCopy(String output) throws IOException, ParserConfigurat
         Document mergeDoc = parser.newDocument();
-        final Comment comment = mergeDoc.createComment(" This config file was generated by pluginUtility merge v1.0.73 on " +
+        final Comment comment = mergeDoc.createComment(" This config file was generated by pluginUtility merge v1.0.75 on " +
                                                        dateTimeFormatter.format(ZonedDateTime.now()) + " ");
```
I'm not sure how this version is determined and why we go from 73 to 75, can you clarify?

emanuel2258 replied:
I am also not certain how they got to the `v1.0.73`. I just changed it to `75` to distinguish between the versions. I can follow another format if you think would be more appropriate but I would not like to leave it with `v1.0.73` with all these code changes as this gets printed out in the plugin-cfg.xml and want to make sure it's easy to see what version it was used to generate the plugin config.xml

emanuel2258 replied:
@pnicolucci ok to resolve this or have any additional comments?

pnicolucci replied:
Why increment by 2 and not just 1? Does the +2 correlate to something?

pnicolucci replied:
My guess is that this change was made in the previous PR that caused this problem and was likely based on the bundle version at that point in time. That bundle version will be updated with every release but this is static so going forward likely the best option is to just increment if you want to.

pnicolucci replied:
So same question , why increment by 2?

volosied replied:
I would use `v1.0.102` as that's the current bundle version (in integration as of today)

pnicolucci replied:
That's fine with me

emanuel2258 replied:
I will increase it by 1

pnicolucci commented on dev/com.ibm.ws.http.plugin.merge/src/com/ibm/ws/http/plugin/merge/internal/PluginMergeToolImpl.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2013, 2023 IBM Corporation and others.
+ * Copyright (c) 2013, 2025 IBM Corporation and others.
```
You can also remove the "contributors" section of the copyright statement. It is no longer needed. I try to clean these up as we touch files.

pnicolucci reviewed (commented):
Cleaning up your commit history would be great, there are commits there with descriptions that don't make sense any longer as the design of the fix has changed since some of the commits were added. Adding a test case as we've discussed would be excellent as well.

emanuel2258 commented:
@pnicolucci I made the requested code changes and cleaned up the commit you mentioned. In terms of the test do you recommend I add it to the testbucket `com.ibm.ws.webserver.plugin.runtime_fat` in closed liberty or creating a new one? It seems this test bucket already covers some test for pluginUtility cluster generate so wonder if I just add to it.

emanuel2258 commented:
I did have a discussion with Eric this morning regarding the code changes and he seemed ok with them. I went ahead and  created a testFix for the customer.

pnicolucci commented:
> @pnicolucci I made the requested code changes and cleaned up the commit you mentioned. In terms of the test do you recommend I add it to the testbucket `com.ibm.ws.webserver.plugin.runtime_fat` in closed liberty or creating a new one? It seems this test bucket already covers some test for pluginUtility cluster generate so wonder if I just add to it.

If we have an existing bucket where it makes sense to add these tests then I have no objection to using that existing FAT project. Let me know if you have any other questions.

emanuel2258 commented:
@pnicolucci as requested, I have added a cached plugin-cfg.xml regeneration test in https://github.ibm.com/websphere/WS-CD-Open/pull/34255

emanuel2258 commented:
@pnicolucci The customer came back saying the testFix worked for them. Is there any remaining items that are necessary for the approval of this PR or can it be approved?

pnicolucci commented on dev/com.ibm.ws.http.plugin.merge/src/com/ibm/ws/http/plugin/merge/internal/PluginMergeToolImpl.java:
```diff
@@ -6,9 +6,6 @@
  * http://www.eclipse.org/legal/epl-2.0/
```
You should update the copyright date range to include 2025

pnicolucci replied:
2013, 2025

emanuel2258 replied:
Ok

emanuel2258 committed:
Changing copyright year and the version of the pluginUtility

pnicolucci commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=73225c4a-d8a3-4bda-93ea-a37f080309a5) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

emanuel2258 commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=d75f916f-e31f-473d-9de6-0829765f90ed) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

emanuel2258 commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=97adf61f-459e-4c37-8a4a-bb21641512d3) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

emanuel2258 commented:
<img width="1031" alt="image" src="https://github.com/user-attachments/assets/4c635c75-2470-4609-8915-f3a3bc7a8918" />
Plugin test looks ok. Will merge the PR.
